data_structures/read_group.wdl

```diff
@@ -0,0 +1,386 @@
+        }
+    }
+
+    parameter_meta {
+        bam: "Input BAM format file to get read groups from"
+        modify_disk_size_gb: "Add to or subtract from dynamic disk space allocation. Default disk size is determined by the size of the inputs. Specified in GB."
+    }
+
+    input {
+        File bam
+        Int modify_disk_size_gb = 0
+    }
+
+    Float bam_size = size(bam, "GiB")
+    Int disk_size_gb = ceil(bam_size) + 10 + modify_disk_size_gb
+
+    command <<<
+        set -euo pipefail
+        BAM="~{bam}" OUTFILE="read_groups.json" python - <<END
+import os  # lint-check: ignore
+import pysam  # lint-check: ignore
+import json  # lint-check: ignore
+sam = pysam.AlignmentFile(os.environ["BAM"], "rb")
+
+out_file = open(os.environ["OUTFILE"], "w")
+json.dump(sam.header.to_dict()["RG"], out_file)
+out_file.close()
+END
+    >>>
+
+    output {
+        Array[ReadGroup] read_groups = read_json("read_groups.json")
+    }
+
+    runtime {
+        memory: "4 GB"
+        disk: "~{disk_size_gb} GB"
+        container: 'quay.io/biocontainers/pysam:0.22.0--py38h15b938a_1'
+        maxRetries: 1
+    }
+}
+
+task validate_ReadGroup {
+    meta {
+        description: "Validate a ReadGroup struct's fields are defined"
+        outputs: {
+            check: "Dummy output to indicate success and enable call-caching"
+        }
+    }
+
+    parameter_meta {
+        read_group: "ReadGroup struct to validate"
+        required_fields: "Array of read group fields that must be defined"
+        restrictive: "If true, run a less permissive validation of field values. Otherwise, check against SAM spec-defined values."
+    }
+
+    input {
+        ReadGroup read_group
+        Array[String] required_fields = ["ID"]
+        Boolean restrictive = true
+    }
+
+    # The SAM spec allows any printable ASCII character in header fields.
+    String sam_spec_pattern = "[\\ -~]+"
+    # We have the opinion that is too permissive for ID and SM.
+    String id_pattern = if restrictive then "id" else sam_spec_pattern
+    String sample_pattern = if restrictive then "sample.?" else sam_spec_pattern
+    Array[String] platforms = ["CAPILLARY", "DNBSEQ", "ELEMENT", "HELICOS", "ILLUMINA", "IONTORRENT", "LS454", "ONT", "PACBIO", "SINGULAR", "SOLID", "ULTIMA"]
+
+    command <<<
+        error=0
+        if [ $(echo "~{sep(" ", required_fields)}" | grep -Ewc "ID") -eq 1 ]
+        then
+            if [ -z "~{read_group.ID}" ]
+            then
+                >&2 echo "ID is required"
+                error=1
+            fi
+        fi
+        if [ "~{defined(read_group.ID)}" == "true" ]
+        then
+            if [[ "~{read_group.ID}" =~ ^~{id_pattern}$ ]]
+            then
+                >&2 echo "ID must match pattern ~{id_pattern}"
+                error=1
+            fi
+        fi
```

Comment:
IMO it's confusing here to act as if there's a situation where `ID` isn't required. I'd prefer to just directly test `[[ "~{read_group.ID}" =~ ^~{id_pattern}$ ]]` and drop all the other if clauses.